doc/design/zarr-support-2.md 

```diff
@@ -0,0 +1,94 @@
+  Deletes subfiles from S3, and updates the treehash accordingly.
+  Requires a list of file paths
+
+* POST /api/dandisets/{...}/versions/{...}/assets/zarr/{zarr_id}/
+  Creates a new Asset that points to this zarr data.
+  Return an asset ID
+
+## Upload flow
+
+TODO the only gotcha is that published zarr files can't be uploaded to/updated/deleted from
+The slowest API request in the upload flow is the `.../batch/complete/` operation, since it needs to update several `.checksum` files in S3.
+I recommend that the CLI dynamically adjust batch size to keep the `.../batch/complete/` operation under 30s.
+
+## Modification flow
+
+TODO this is basically the same as the upload flow, but with deletes allowed.
+
+## Hashing scheme
+
+To avoid having to store a checksum for every directory in every zarr file, we will delegate the storage of the checksums to S3.
+Update times will be substantially slower since each level of the tree needs to be updated.
+Storage costs will be offloaded to S3, and the treehash will be conveniently accessible to clients.
+
+Let's say the zarr files are kept under `/zarr/{zarr_id}/...` in S3.
+A subfile `1/2/3/4.txt` of zarr file `foobar` would be kept at `/zarr/foobar/1/2/3/4.txt`.
+The ETag can be trivially determined by a `HEAD` request to that S3 object.
+
+After this subfile is uploaded, the API server creates a file `/zarr_checksums/foobar/1/2/3/.checksum` if it doesn't already exist (note the altered prefix).
+This file contains a list of all the files/directories currently in the directory, their checksums, and an aggregated checksum of the entire directory contents.
+If `.checksum` already exists, it is updated with the new file+ETag and the final checksum is updated.
+
+The API server then recursively travels up another level and creates `/zarr_checksums/foobar/1/2/.checksum`.
+This `.checksum` will contain an entry for the directory `3` and the final checksum from `3/.checksum`.
+This recursion continues all the way to the top, yielding a `/zarr/checksums/foobar/.checksum` file which contains an aggregated checksum for the entire zarr file.
+
+### .checksum file format
+
+TODO
+TODO include sha256 as it is calculated? are we even calculating sha256 for zarr files?
+
+### Treehash algorithm
+
+TODO it obviously needs to include all the checksums for all the files in the directory, but assembled how?
+
+## Publishing
+
+When a dandiset containing a zarr file is published, normal Assets are simply added to the new published dandiset, which is a DB only operation and therefore very fast.
+This doesn't work for zarr files, since we want to make sure that subsequent modifications to the draft zarr file don't impact the published zarr file.
+
+Fortunately, fast publishing is not a requirement for dandisets, just a happy coincidedence of our deduplication scheme.
+On publish, simply trigger a job that copies the zarr file S3 data to a new location (TODO where is that exactly?).
```

Comment:
would require locking of that 'zarr file' from modifications interim. copying could take hours or days... also would blow the size of archive twice (or more if multiple versions published with e.g. tiny changes to 'zarr file' content) unnecessarily

I really think we should look into utilizing S3 versioning + external "manifest" file with some custom backend for zarr or may be somehow `fsspec_reference_maker` (e.g. as used [here](https://medium.com/pangeo/fake-it-until-you-make-it-reading-goes-netcdf4-data-on-aws-s3-as-zarr-for-rapid-data-access-61e33f8fe685) for indexing access to .nc files).

Reply:
Good point about locking. I think we could add an error message if clients try to use `{zarr_file}/batch` operations while a publish is happening, which would effectively lock them out of any modifications.

Yep, lots of tiny publishes would be bad. I don't see any way to store the file accessibly in S3 without copying on publish. The alternative would involve a custom zarr backend. I delegate designing that to @yarikoptic